sw/device/tests/usbdev_mem_test.c

```diff
@@ -0,0 +1,212 @@
+                      size_t n) {
+  if (use_dif) {
+    mem_dif_write(dev, data, n);
+  } else {
+    mmio_region_memcpy_to_mmio32(dev->base_addr, USBDEV_BUFFER_REG_OFFSET, data,
+                                 n);
+  }
+}
+
+bool test_main(void) {
+  LOG_INFO("Running USBDEV Mem Test");
+
+  // Initialize DIF-based access to the packet buffer memory
+  CHECK_DIF_OK(
+      dif_usbdev_init(mmio_region_from_addr(USBDEV_BASE_ADDR), &usbdev));
+
+  // We need not be concerned with the device configuration but we DO need the
+  // buffer pool
+  dif_usbdev_config_t config = {
+      .have_differential_receiver = dif_bool_to_toggle(false),
+      .use_tx_d_se0 = dif_bool_to_toggle(false),
+      .single_bit_eop = kDifToggleDisabled,
+      .pin_flip = dif_bool_to_toggle(false),
+      .clock_sync_signals = kDifToggleEnabled,
+  };
+  CHECK_DIF_OK(dif_usbdev_configure(&usbdev, &buffer_pool, config));
+
+  // Exercise mmio_ routines and then dif_usbdev_ routines in turn
+  for (unsigned use_dif = 0u; use_dif <= 1u; use_dif++) {
+    // Set up test pattern
+    uint8_t lfsr = 0x13u;
+    for (unsigned idx = 0u; idx < sizeof(testpatt); idx++) {
+      testpatt[idx] = lfsr;
+      lfsr = LFSR_ADVANCE(lfsr);
+     }
+
+    // Write the generated test pattern to the usbdev packet memory
+    mem_write(&usbdev, use_dif, testpatt, sizeof(testpatt));
+
+    // Invalidate the contents of the copy buffer
+    for (unsigned idx = 0u; idx < sizeof(testpatt); idx++) {
+      testcopy[idx] = 0xffu;
+    }
+
+    // Read back a copy of the test pattern from the usbdev packet memory
+    mem_read(&usbdev, use_dif, testcopy, sizeof(testcopy));
+
+    // Check the copy against the original
+    for (unsigned idx = 0u; idx < sizeof(testpatt); idx++) {
+      if (testcopy[idx] != testpatt[idx]) {
+        LOG_INFO("**** MISMATCH %02x %02x", testpatt[idx], testcopy[idx]);
+      } else if (verbose) {
+        LOG_INFO("patt 0x%02x copy 0x%02x", testpatt[idx], testcopy[idx]);
+      }
+      CHECK(testcopy[idx] == testpatt[idx]);
+
+      // Invert the test pattern to ensure every byte is modified
+      testcopy[idx] = ~testcopy[idx];
+    }
+
+    // Write the modified test pattern to the device memory
+    mem_write(&usbdev, use_dif, testcopy, sizeof(testcopy));
+
+    // Invalidate the original test pattern buffer
+    for (unsigned idx = 0u; idx < sizeof(testpatt); idx++) {
+      testpatt[idx] = 0xffu;
+    }
+
+    // Read back the modified test pattern from the device memory
+    mem_read(&usbdev, use_dif, testpatt, sizeof(testpatt));
+
+    // Check against the modified test pattern
+    for (unsigned idx = 0u; idx < sizeof(testpatt); idx++) {
+      if (testcopy[idx] != testpatt[idx]) {
+        LOG_INFO("**** MISMATCH %02x %02x", testpatt[idx], testcopy[idx]);
+      } else if (verbose) {
+        LOG_INFO("copy 0x%02x patt 0x%02x", testcopy[idx], testpatt[idx]);
+      }
+      CHECK(testcopy[idx] == testpatt[idx]);
```

Comment:
Maybe you want:
```suggestion
      CHECK(testcopy[idx] == testpatt[idx], "copy 0x%02x patt 0x%02x", testcopy[idx], testpatt[idx]);
```
Or maybe replace the whole `for` loop by the `CHECK_ARRAYS_EQ` macro, which will log mismatches for you.

Reply:
Thanks very much. I didn't know that macro existed!